packages/language-server/src/completion-items.ts

```diff
@@ -190,13 +256,63 @@ function getClassNamespacePrefix(
   return undefined;
 }
 
+function getNodeFilterText(suggestion: UI5XMLViewCompletion): string {
+  // The filter text is used by VSCode/Theia to filter out suggestions that don't match the text the user wrote.
+  // Basically this means that every character being replaced by the TextEdit should exist in the filter text.
+  // Since we replace the whole value (tag name/attribute key/attribute value) the filter text should contain
+  // everything which should be replaced (including namespace in tag name, surrounding quotation marks etc).
+  // By default we replace the simple name of the node so this is the returned filter text.
+  // Special cases:
+  // - Class name in tag can be filtered by FQN or xmlns:
+  //   For sap.f.Cards, user input "f:C" should return "f:Cards"
+  //   For sap.m.Button, user input "sap.m.But" should return "sap.m.Button"
+  // - Namespace in attribute value can be filtered by FQN
+  //   For sap.m.semantic, user input "sap.m." should return "sap.m.semantic"
+  // - Namespace in xmlns attribute key should contain the "xmlns:" prefix
+  //   For sap.m, user input "xmlns:" should return "xmlns:m"
+  // - Attribute values should contain quotation marks
+  //   For Small enum value, user input `"S"` should return `"Small"`
+  const node = suggestion.ui5Node;
+  switch (suggestion.type) {
+    case "UI5ClassesInXMLTagName": {
+      // In some cases ns will have the namespace and in others name will contain the namespace.
+      // If the user explicitly typed a namespace we use it in the filter text (because it's written in the file).
+      if (
+        !isEmpty(suggestion.astNode.ns) ||
+        (suggestion.astNode.name !== null &&
+          includes(suggestion.astNode.name, ":"))
+      ) {
+        const prefix = getClassNamespacePrefix(suggestion);
+        // The else case should never happen - if the user entered a namespace in the tag name only classes from
+        // that namespace are returned, so the namespace will be returned in the prefix
+        /* istanbul ignore else */
+        if (prefix !== undefined) {
```

Comment:
should we put the check `getClassNamespacePrefix(suggestion) !== undefined` as part of the conditions in lines 281-283 ?

I know it means we will call it twice, but maybe that will allow the typescript compiler to understand that it is never undefined?

Reply:
I can do the assignment in the `if` so we won't call it twice... not sure it will be better. It would be less readable. I prefer to keep the ignore. (Also it's not a compilation issue, it's a logical issue with a reasonable fallback in case we have a bug).